### ISSUE TITLE
(SSL routines', 'SSL3_WRITE_PENDING', 'bad write retry')

### ISSUE THREAD

**Issue by jlvcm** [Azure/azure-sdk-for-python#499]
when using put_block_blob_from_text
i get this error:
```
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/blob/blobservice.py", line 1352, in put_block_blob_from_text
    retry_wait)
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/blob/blobservice.py", line 1228, in put_block_blob_from_bytes
    x_ms_lease_id)
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/blob/blobservice.py", line 883, in put_blob
    self._perform_request(request)
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/storageclient.py", line 171, in _perform_request
    resp = self._filter(request)
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/storageclient.py", line 160, in _perform_request_worker
    return self._httpclient.perform_request(request)
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/_http/httpclient.py", line 181, in perform_request
    self.send_request_body(connection, request.body)
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/_http/httpclient.py", line 143, in send_request_body
    connection.send(request_body)
  File "/root/appuro-env/lib/python2.7/site-packages/azure/storage/_http/requestsclient.py", line 81, in send
    self.response = self.session.request(self.method, self.uri, data=request_body, headers=self.headers, timeout=self.timeout)
  File "/root/appuro-env/lib/python2.7/site-packages/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/root/appuro-env/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/root/appuro-env/lib/python2.7/site-packages/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/root/appuro-env/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 559, in urlopen
    body=body, headers=headers)
  File "/root/appuro-env/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 353, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib64/python2.7/httplib.py", line 973, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib64/python2.7/httplib.py", line 1007, in _send_request
    self.endheaders(body)
  File "/usr/lib64/python2.7/httplib.py", line 969, in endheaders
    self._send_output(message_body)
  File "/usr/lib64/python2.7/httplib.py", line 829, in _send_output
    self.send(msg)
  File "/usr/lib64/python2.7/httplib.py", line 805, in send
    self.sock.sendall(data)
  File "/root/appuro-env/lib/python2.7/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 220, in sendall
    sent = self._send_until_done(data[total_sent:total_sent+SSL_WRITE_BLOCKSIZE])
  File "/root/appuro-env/lib/python2.7/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 206, in _send_until_done
    return self.connection.send(data)
  File "/root/appuro-env/lib/python2.7/site-packages/OpenSSL/SSL.py", line 1271, in send
    self._raise_ssl_error(self._ssl, result)
  File "/root/appuro-env/lib/python2.7/site-packages/OpenSSL/SSL.py", line 1187, in _raise_ssl_error
    _raise_current_error()
  File "/root/appuro-env/lib/python2.7/site-packages/OpenSSL/_util.py", line 48, in exception_from_error_queue
    raise exception_type(errors)
Error: [('SSL routines', 'SSL3_WRITE_PENDING', 'bad write retry')]
```

i've uninstalled ndg-httpsclient to temporaly "fix" it

Python 2.7.5
azure==1.0.2
azure-common==1.0.0
azure-mgmt==0.20.1
azure-mgmt-common==0.20.0
azure-mgmt-compute==0.20.0
azure-mgmt-network==0.20.1
azure-mgmt-nspkg==1.0.0
azure-mgmt-resource==0.20.1
azure-mgmt-storage==0.20.0
azure-nspkg==1.0.0
azure-servicebus==0.20.1
azure-servicemanagement-legacy==0.20.1
azure-storage==0.20.2
requests==2.8.1
requests-oauthlib==0.5.0
pyOpenSSL==0.15.1

**Comment by huguesv** [comment]
Thanks for the report.  I've tried it and wasn't able to repro the issue on my machine (win8.1).

It would be good to know your operating system distro+version.  I don't know that there is anything that we can change in the library to avoid this issue, but we could try upgrading your distro's Python to the latest 2.7.x and see if that helps.

```
2.7.9 (default, Dec 10 2014, 12:24:55) [MSC v.1500 32 bit (Intel)]

azure-common==1.0.0
azure-nspkg==1.0.0
azure-storage==0.20.3
cffi==1.3.0
cryptography==1.1
enum34==1.0.4
futures==3.0.3
idna==2.0
ipaddress==1.0.15
ndg-httpsclient==0.4.0
pyasn1==0.1.9
pycparser==2.14
pyOpenSSL==0.15.1
python-dateutil==2.4.2
requests==2.8.1
six==1.10.0
```

**Comment by jlvcm** [comment]
My OS/distro:
CentOS 7 (Core) - 7.1.1503

Enabled repos (it's fully updated):
- CentOS7 - Base
  - http://ftp.dei.uc.pt/pub/linux/CentOS/7.1.1503/os/x86_64/
- Extra Packages for Enterprise Linux 7
  - http://ftp.up.pt/fedora-epel/7/x86_64/
- CentOS-7 - Extras
  - http://ftp.dei.uc.pt/pub/linux/CentOS/7.1.1503/extras/x86_64/
- RHEL 7 - RPMforge.net - dag
  - http://apt.sw.be/redhat/el7/en/x86_64/rpmforge/
- CentOS-7 - Updates
  - http://ftp.dei.uc.pt/pub/linux/CentOS/7.1.1503/updates/x86_64/



i'm using python 2.7.5 as it is my distro default version, but I've tested with 2.7.10 and i get the same issue

**Comment by lmazuel** [comment]
Hi,

You bug belongs to the storage library, and there is already a bug opened:
https://github.com/Azure/azure-storage-python/issues/54

But it seems to be a bug in `urllib3` embedded in `requests` 2.8.1. Since now `requests` 2.9.1 is released with `urllib` 1.13.1 inside, could you confirm that your problem is solved?

Thank you,

**Comment by lmazuel** [comment]
Hi,

The related bug in the Azure storage github account was reported fixed by updating to `requests` 2.9.1.
I close this bug, but feel free to re-open it if you think it's not fixed in your case.

Thank you!

**Comment by lavalamp-** [comment]
It looks like this bug may have been re-introduced in version 2.13.0. I was struggling quite a bit to fix this (working with Elasticsearch's requests adapter) and downgrading to 2.11.1 fixed the problem for me.
